### PR TITLE
Added __getattr__ to VecEnvWrapper

### LINKED ISSUE #274: Improve __getattr__ for VecEnvWrapper
In more complicated experiments, a user may wish to wrap a VecEnv with multiple instances of `VecEnvWrapper` in a modular fashion. For example, we could have something like below:

```python
class RewardVecWrapper(VecEnvWrapper):
    def __init__(self, venv, reward_multiplier):
        super().__init__(venv)
        self._reward_multiplier = reward_multiplier

    def step_wait(self):
        obs, rewards, dones, infos = self.venv.step_wait()
        rewards *= self._reward_multiplier
        return obs, rewards, dones, infos

    def get_reward_multiplier(self):
        return self._reward_multiplier

class LoggingVecWrapper(VecEnvWrapper):
     def custom_log(self, infos):
           # stateful logging, saving to file in custom format, etc. 
           ...

     def step_wait(self):
         obs, rewards, dones, infos = self.venv.step_wait()
         self.custom_log(infos)
         return obs, rewards, dones, infos
```

The issue arises if I have a venv which has been wrapped with `RewardVecWrapper` and and I wish to call `get_reward_multiplier`. Depending on if I have wrapped it with `LoggingVecWrapper` (or any number of wrappers really), I would have to do some unappealing type-checking, something to the effect of 

```python
def get_arbitrarily_deep_reward_multiplier(wrapped_venv):
    reward_venv = wrapped_venv
    while True:
        if isinstance(reward_venv, RewardVecWrapper):
            return reward_venv.get_reward_multiplier()
        else:
            reward_venv = reward_venv.venv
```

I believe we can fix this by modifying `VecEnvWrapper.__getattr__` to the following: 
```python
# in class VecEnvWrapper
def __getattr__(self, name):
    # Edit: line below was originally "return self.venv.__getattr__(name)"
    return getattr(self.venv, name)
```

This way, as long as different VecEnvWrappers have unique names for attributes of interest, one can always find that attribute from a top-level call to `venv.__getattr__`.

### PR DESCRIPTION
**Pull request by kantneel**
Closes #274 .

### CODE REVIEW

**Comment by kantneel** [comment]
@AdamGleave @hill-a @araffin I've implemented the check for unique attribute references, added a test and updated the changelog. Please review these changes. Also, I'm not sure why the codacy check failed on the penultimate commit; It seems fine when I click on it.

**Comment by AdamGleave** [comment]
You had a bug that `getattr_recursive` is never in `self.__dict__`, so I changed it to use `hasattr`. Also cleaned up the logic a bit: you can do the depth check directly in `__getattr__` which avoids the need for the `top_level` logic. And added a few tests.

I'm mostly happy with this PR. The one thing I'm uncomfortable about is that `__getattr__` will *only* raise when a child wrapper shadows a variable in another child, not when the current wrapper shadows a child variable. This is because `__getattr__` won't even get called when something is present in `self.__dict__`. I can see the argument for this: there are some things we do actually want to shadow (e.g. `step_wait`), and it'd be unfortunate if there was not a way to access some attributes at all because of a name clash. But it does make the whole thing seem confusing to me. This should perhaps be stressed more in the documentation?

I'm not comfortable reviewing now I've touched the code so would appreciate if @hill-a or @araffin could take a look.

**Comment by araffin** [comment]
I'll try to do the review tomorrow, ping me if I don't ;)

File: stable_baselines/common/vec_env/base_vec_env.py [hunk 1]
```diff
@@ -150,6 +150,19 @@ def unwrapped(self):
         else:
             return self
 
+    def getattr_depth_check(self, name, already_found):
+        """Check if an attribute reference is being hidden in a recursive call to __getattr__
+
+        :param name (str): name of attribute to check for
+        :param already_found (bool): whether this attribute has already been found in a wrapper
+        :return: (str or None) blocked_venv_class
```
Comment (araffin) [inline stable_baselines/common/vec_env/base_vec_env.py]: either missing comment or it needs a better name
I'm not sure what `blocked_venv_class` means.

File: stable_baselines/common/vec_env/base_vec_env.py [hunk 2]
```diff
@@ -207,6 +220,54 @@ def set_attr(self, attr_name, value, indices=None):
...
 
+    def __getattr__(self, name):
+        """Find attribute from wrapped venv(s) if this wrapper does not have it.
+        Useful for accessing attributes from venvs which are wrapped with multiple wrappers
+        which have unique attributes of interest.
+        """
+        blocked_class = self.getattr_depth_check(name, already_found=False)
+        if blocked_class is not None:
+            own_class = "{0}.{1}".format(type(self).__module__, type(self).__name__)
+            fstr = ("Error: Recursive attribute lookup for {0} from {1} is"
```
Comment (araffin) [inline stable_baselines/common/vec_env/base_vec_env.py]: `formatted_str` would be a better name ? or at least `f_str`.

**Comment by AdamGleave** [comment]
Thanks for the review @araffin, I've addressed those comments.

**Pull request merged**